lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -204,23 +214,103 @@ export abstract class CacheManager implements ICacheManager {
+                const entity: AccountEntity | null = this.getAccount(key);
+                if (!entity) {
+                    return tenantProfileAccounts;
+                }
+                tenantProfileAccounts.push(this.getAccountInfoFromEntity(entity));
+                return tenantProfileAccounts;
+            }, []);
+            return this.buildMultiTenantAccount(baseAccount, tenantProfileAccounts);
+        });
+    }
 
-    /** 
-     * Gets accountInfo object based on provided filters
+    /**
+     * Finds all cached accounts that are tenant profiles of the base account and
+     * populates the base account's tenant profile map with their profile data
+     * @param baseAccount 
+     * @returns 
      */
-    getAccountInfoFilteredBy(accountFilter: AccountFilter): AccountInfo | null{
-        const allAccounts = this.getAccountsFilteredBy(accountFilter);
-        if (allAccounts.length > 0) {
-            return this.getAccountInfoFromEntity(allAccounts[0]);
+    buildMultiTenantAccount(baseAccount: AccountInfo, preloadedTenantProfileAccounts?: AccountInfo[]): AccountInfo {
+        const tenantProfiles = new Map<String, TenantProfile>();
+        const accountId = [baseAccount.homeAccountId, baseAccount.environment].join(Separators.CACHE_KEY_SEPARATOR);
+        let tenantProfileAccounts: AccountInfo[] = [];
+
+        // For cases where we already have the tenant profiles preloaded like getAllAccountsMultiTenant, tenantProfiles should be passed in to avoid repeated cache access
+        if (preloadedTenantProfileAccounts) {
+            tenantProfileAccounts = preloadedTenantProfileAccounts;
         } else {
-            return null;
+            // Assuming this method is called for one account in isolation, cache access perf hit is acceptable
+            const cachedAccountKeys = this.getAccountKeys();
+            // Filter accont keys belonging to user accross tenants (including home tenant)
+            const tenantProfileAccountKeys = cachedAccountKeys.filter((key:string) => {
+                return key.indexOf(accountId) === 0;
+            });
+
+            tenantProfileAccountKeys.forEach((key: string) => {
+                const tenantProfileAccount = this.getAccount(key)?.getAccountInfo();
+                if (tenantProfileAccount) {
+                    tenantProfileAccounts.push(tenantProfileAccount);
+                }
+            });
         }
+
+        // Map tenant profile accounts to their tenantProfiles for baseAccount
+        tenantProfileAccounts.forEach((tenantProfileAccount: AccountInfo) => {
+            tenantProfiles.set(
+                tenantProfileAccount.tenantId, { 
+                    objectId: tenantProfileAccount.localAccountId,
+                    tenantId: tenantProfileAccount.tenantId,
+                    isHomeTenant: tenantProfileAccount.tenantId === tenantProfileAccount.homeAccountId.split(".")[1]
```

Comment:
Need to make sure any whitespace on either end is stripped off before comparison